main/src/cgeo/geocaching/files/GPXImporter.java

```diff
@@ -167,91 +165,101 @@ public void importGPX() {
     }
 
     private static final class ProgressHandler extends DisposableHandler {
-        private final WeakReference<GPXImporter> importerRef;
+        private final WeakReference<Progress> progressRef;
 
-        ProgressHandler(final GPXImporter importer) {
-            importerRef = new WeakReference<>(importer);
+        ProgressHandler(final Progress progress) {
+            progressRef = new WeakReference<>(progress);
         }
 
         @Override
         public void handleRegularMessage(final Message msg) {
-            final GPXImporter importer = importerRef.get();
-            if (importer != null) {
-                importer.progress.setProgress(msg.arg1);
+            final Progress progress = progressRef.get();
+            if (progress != null) {
+                progress.setProgress(msg.arg1);
             }
         }
     }
 
-    private static final class ImportStepHandler extends WeakReferenceHandler<GPXImporter> {
+    private static final class ImportStepHandler extends Handler {
+        private final Activity fromActivity;
+        private final Handler importFinishedHandler;
+        private final Progress progress;
+        private final WeakReference<DisposableHandler> progressHandlerRef;
+        private final Resources res;
+
+        ImportStepHandler(final GPXImporter importer, final Activity fromActivity) {
+            this.fromActivity = importer.fromActivity;
+            this.importFinishedHandler = importer.importFinishedHandler;
+            this.progress = importer.progress;
+            progressHandlerRef = new WeakReference<>(importer.progressHandler);
+            res = fromActivity.getResources();
+        }
 
-        ImportStepHandler(final GPXImporter importer) {
-            super(importer);
+        private void disposeProgressHandler() {
+            final DisposableHandler progressHandler = progressHandlerRef.get();
+            if (progressHandler != null) {
+                progressHandler.dispose();
+            }
+        }
+
+        private void importFinished() {
+            if (importFinishedHandler != null) {
+                importFinishedHandler.sendEmptyMessage(0);
+            }
         }
 
         @Override
         public void handleMessage(final Message msg) {
-            final GPXImporter importer = getReference();
-            if (importer != null) {
```

Comment:
Please prefer guards over nested control flow structures. However, there is no need to change this for the PR.